pathcore/network_permutation_test.py

```diff
@@ -0,0 +1,185 @@
+        permuted_network = pathcore_network.permute_pathways_across_features()
+        if not aggregate_permuted_network:
+            aggregate_permuted_network = permuted_network
+        else:
+            aggregate_permuted_network.aggregate(permuted_network)
+    return aggregate_permuted_network
+
+
+def network_edges_permutation_test(observed_network,
+                                   permuted_networks,
+                                   alpha,
+                                   n_networks=1,
+                                   output_edges_to_file=None,
+                                   output_network_to_file=None):
+    """Given the observed network and N permutations of that network,
+    determine the significance of each observed, weighted edge. Edges that
+    are not distinguishable from the null model of random associations will
+    be removed from the network file output. Additionally, weights of the
+    significant edges are updated to an odds ratio that quantifies the
+    "unexpectedness" of an observed pathway co-occurrence relationship
+    based on the generated null model.
+
+    Parameters
+    -----------
+    observed_network : CoNetwork
+      the original observed network (this can be an aggregate of
+      multiple networks, see CoNetwork's `aggregate` function in
+      `network.py`)
+    permuted_networks : list(CoNetwork)
+      the list of N permuted networks generated from the original observed
+      network
+    alpha : float
+      specify the threshold for significance testing
+    n_networks : int (default=1)
+      in the case where `observed_network` is an aggregate of multiple
+      co-occurrence networks, `n_networks` can be greater than 1.
+    output_edges_to_file : str|None (default=None)
+      specify the filepath to write the edges that had q-values
+      below `alpha`
+    output_network_to_file : str|None (default=None)
+      specify the filepath to write the final, filtered network
+
+    Returns
+    -----------
+    CoNetwork, the observed network after the permutation test (filtered
+    to only report significant edges, weighted by their odds ratios)
+    """
+    n_permutations = len(permuted_networks)
+    n_features = observed_network.n_features
+
+    edge_weight_distributions = _get_edge_weight_distributions(
+        observed_network, permuted_networks, n_networks * n_features)
+
+    pvalues = []
+    edge_expected_weight = []
+    for edge_id, edge_obj in observed_network.edges.items():
+        observed_weight = edge_obj.weight
+        weight_distribution = edge_weight_distributions[edge_id]
+        eq_or_above_observed = sum(weight_distribution[observed_weight:])
+        pvalue = eq_or_above_observed / float(n_permutations)
+        pvalues.append(pvalue)
+        edge_expected_weight.append((edge_id, _edge_expected_weight(
+            weight_distribution[1:])))
+
+    below_alpha, qvalues, _, _ = multipletests(
+        pvalues, alpha=alpha, method="fdr_bh")
```

Comment:
false discovery rate, benjamini & hochberg. will add as a comment above this line